# Main entrypoint of the workflow.
# Please follow the best practices:
# https://snakemake.readthedocs.io/en/stable/snakefiles/best_practices.html,
# in particular regarding the standardized folder structure mentioned there.

import glob
import os

configfile: "config/config.yml"

input_dir = config["input_dir"]
output_dir = config["output_dir"]

# Collect sample names from test_data directory
SAMPLES = [os.path.basename(d) for d in glob.glob(f"{input_dir}/*") if os.path.isdir(d)]

rule all:
    input:
        expand(f"{output_dir}" + "/{sample}/spades/scaffolds.fasta", sample=SAMPLES),
        expand(f"{output_dir}" + "/{sample}/megahit", sample=SAMPLES),
        expand(f"{output_dir}" + "/{sample}/abyss/abyss-scaffolds.fa", sample=SAMPLES),
        expand(f"{output_dir}" + "/{sample}/sparseassembler/SuperContigs.txt", sample=SAMPLES),
        expand(f"{output_dir}" + "/{sample}/minia/{sample}.contigs.fa", sample=SAMPLES),
#        expand(f"{output_dir}" + "/{sample}/masurca/masurca_config.txt", sample=SAMPLES),
        expand(f"{output_dir}" + "/{sample}/masurca/CA/primary.genome.scf.fasta", sample=SAMPLES),

# include spades rules
include: "rules/spades.smk"
# include megahit rule
include: "rules/megahit.smk"
# include abyss rule
include: "rules/abyss.smk"
# include decompress_reads rule
include: "rules/decompress_reads.smk"
# include sparseassembler rule
include: "rules/sparseassembler.smk"
# include minia rule
include: "rules/minia.smk"
# include masurca_config rule
include: "rules/masurca_config.smk"
# include masurca rule
include: "rules/masurca.smk"

#print("SAMPLES:", SAMPLES)
